File: Python/138. Copy List with Random Pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        nodes = {}

        def dfs(node):
            if not node: return
            if node in nodes: return nodes[node]

            new = Node(node.val)
            nodes[node] = new

            new.next = dfs(node.next)
            new.random = dfs(node.random)

            return new

        return dfs(head)

File: Java/138. Copy List with Random Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public HashMap<Node, Node> nodes = new HashMap<Node, Node>();

    public Node dfs(Node node) {
            if (node == null) return null;
            if (nodes.containsKey(node)) return nodes.get(node);

            Node n = new Node(node.val);
            nodes.put(node, n);

            n.next = dfs(node.next);
            n.random = dfs(node.random);

            return n;
        }

    public Node copyRandomList(Node head) {
        return dfs(head);
    }
}
